azext_iot/monitor/property.py

```diff
@@ -7,91 +7,202 @@
+        last_updated = isodate.parse_datetime(val["$lastUpdated"])
+        return last_updated.timestamp() >= updated_within.timestamp()
+
+    def _changed_props(self, prop, metadata, property_name):
+        # not an interface - whole thing is change log
+        if not self._is_interface(property_name):
+            return prop
+        # iterate over property in the interface
+        # if the property is not an exact match for what is present in the previous set of properties
+        # track it as a change
+        diff = {
+            key: prop[key]
+            for key, val in metadata.items()
+            if self._is_relevant(key, val)
+        }
+        return diff
+
+    def _validate_payload(self, changes, minimum_severity):
+        for value in changes:
+            self._validate_payload_against_interfaces(changes[value], value)
+
+        issues = self._issues_handler.get_issues_with_minimum_severity(minimum_severity)
+
+        for issue in issues:
+            issue.log()
+
+    def _validate_payload_against_interfaces(self, payload: dict, name):
+        name_miss = []
+        interface_name = name.replace(PNP_INTERFACE_PREFIX, "")
+        if self._is_interface(interface_name):
+            # if the payload is an interface then iterate thru the properties under the interface
+            for property_name in payload.keys():
+                schema = self._template.get_schema(
+                    name=property_name, interface_name=interface_name
+                )
+                if not schema:
+                    name_miss.append(property_name)
+        else:
+            # if the payload is a property then process the payload as a single unit.
+            schema = self._template.get_schema(name=name, interface_name="")
+
+            if not schema:
+                name_miss.append(name)
+
+            if self._validate_duplicate_properties(name):
+                details = strings.duplicate_property_name(
+                    name, list(self._template.interfaces.keys())
+                )
+                self._add_central_issue(severity=Severity.warning, details=details)
+
+        if name_miss:
+            details = strings.invalid_field_name_mismatch_template(
+                name_miss, self._template.schema_names
             )
+            self._add_central_issue(severity=Severity.warning, details=details)
 
-            if change_d:
-                print("Changes in desired properties:")
-                print("version :", twin.desired_property.version)
-                print(change_d)
-
-            if change_r:
-                print("Changes in reported properties:")
-                print("version :", twin.reported_property.version)
-                print(change_r)
+    def _validate_duplicate_properties(self, property_name):
+        value = (
+            sum(
+                property_name in idnumber
+                for idnumber in self._template.interfaces.values()
```

Comment:
I'd even argue that this "find interfaces that have a schema with name matching" is useful enough to include in the template class

Reply:
This will also remove a minor bug in the code. 

Currently here: https://github.com/Azure/azure-iot-cli-extension/pull/212/files#diff-209395f6b9ec07ce7b03dd88c1122426R105

If a duplicate property is detected, the code will print out all interfaces, rather than just the interfaces that have the duplicate property.

e.g.:

```
interfaces = {'a': 'bool', 'b': 'bool', 'c': 'date'}
dups = _validate_duplicate_properties('bool')
if dups: print...
```

the print result will be:

['a', 'b', 'c']

rather than

['a', 'b']